### PR TITLE
ARROW-62: Clarify null bitmap interpretation, indicate bit-endianness, add null count, remove non-nullable physical distinction

### PR DESCRIPTION
**Pull request by wesm**
As the initial scribe for the Arrow format, I made a mistake in what the null bits mean (1 for not-null, 0 for null). I also addressed ARROW-56 (bit-numbering) here.

Database systems are split on this subject. PostgreSQL for example does it this way:

http://www.postgresql.org/docs/9.5/static/storage-page-layout.html

> In this list of bits, a 1 bit indicates not-null, a 0 bit is a null. When the bitmap is not present, all columns are assumed not-null.

Since the Drill implementation predates the Arrow project, I think it's safe to go with this. 

This patch also includes ARROW-76 which adds a "null count" to the memory layout indicating the actual number of nulls in an array. This also strikes the "non-nullable" distinction from the memory layout as there is no semantic difference between arrays with null count 0 and a non-nullable array. Instead, users may choose to set `nullable=false` in the schema metadata and verify that Arrow memory conforms to the schema.

### CODE REVIEW

File: format/Layout.md [hunk 1]
```diff
@@ -90,11 +90,27 @@ maximum of 2^31 - 1 elements. We choose a signed int32 for a couple reasons:
 Any relative type can be nullable or non-nullable.
 
 Nullable arrays have a contiguous memory buffer, known as the null bitmask,
```
Comment (emkornfield) [inline format/Layout.md]: It might be better to frame this paragraph in terms of validity.  Drills documentation phrases it well I think (it doesn't reference a null bitmark)

> Nullable values are represented by a vector of bit values. Each bit in the vector corresponds to an element in the ValueVector. If the bit is not set, the value is NULL.

**Comment by emkornfield** [comment]
If you do decide rephrase the bitmask as a validity vector, you should probably update the documentation  on the flatbufferschema

> 
> /// The number of buffers appended to this list depends on the schema. For
>   /// example, most primitive arrays will have 2 buffers, 1 for the null bitmap
>   /// and 1 for the values. For struct arrays, there will only be a single
>   /// buffer for the null bitmap

**Comment by wesm** [comment]
I rephrased the language a little bit. I'll appeal to one of the other committers to review.

**Comment by wesm** [comment]
Postgres uses the term "null bitmap" if that seems reasonable so I will try to use that consistently in the code and format docs

**Comment by emkornfield** [comment]
Thanks, I probably should have prefaced the above with IMHO.  In my mind, I prefix the name of the bitmap with an "is_" and assume 1 means true.

File: format/Layout.md [hunk 2]
```diff
@@ -89,8 +89,9 @@ maximum of 2^31 - 1 elements. We choose a signed int32 for a couple reasons:
 
 Any relative type can be nullable or non-nullable.
 
-Nullable arrays have a contiguous memory buffer, known as the validity (or
-null) bitmap, whose length is large enough to have 1 bit for each array slot.
+Nullable arrays have a contiguous memory buffer, known as the null (or
+validity) bitmap, whose length is large enough to have 1 bit for each array
+slot.
```
Comment (jacques-n) [inline format/Layout.md]: I would propose that the null bitmap is always an multiple of 8 bytes in length. This simplifies some code to avoid having to manage partial word conditions.

File: format/Layout.md [hunk 3]
```diff
@@ -113,9 +114,9 @@ j mod 8   7  6  5  4  3  2  1  0
           0  0  1  0  1  0  1  1
 ```
 
-Physically, non-nullable (NN) arrays do not have a bitmap.
+Physically, non-nullable (NN) arrays do not have a null bitmap.
```
Comment (jacques-n) [inline format/Layout.md]: I thought our thinking was to avoid the concept of non-nullable arrays?

File: format/Layout.md (hunk 2 above)
Comment (wesm) [inline format/Layout.md]: I agree. There's also the SIMD question — if these buffers are word-aligned then there won't be concerns (? someone with more expertise should opine) with aligned allocations

File: format/Layout.md (hunk 3 above)
Comment (wesm) [inline format/Layout.md]: Totally. I'm going to make another patch to strike nullability from the memory layout, but didn't want to pack too many changes into this patch https://issues.apache.org/jira/browse/ARROW-76

**Comment by wesm** [comment]
@jacques-n I'm going to go ahead and expand the patch to account for ARROW-76, changes to be posted shortly. I'll await your +1 and further comments.

**Comment by wesm** [comment]
Merging these format changes. Debate on these subjects may continue on the mailing list. Thank you
